overlord/devicestate/devicestate_install_api_test.go

```diff
@@ -0,0 +1,439 @@
+		return nil, nil
+	})
+	s.AddCleanup(restore)
+
+	// Note that ESP must be mounted in the same place as a seed partition
+	// so MarkRecoveryCapableSystem is happy when searching for a bootloader.
+	// TODO Should this be changed?
+	espDir := filepath.Join(dirs.RunDir, "mnt/ubuntu-seed")
+
+	// Mock mounting of partitions
+	mountVolsCalls := 0
+	restore = devicestate.MockInstallMountVolumes(func(onVolumes map[string]*gadget.Volume, encSetupData *install.EncryptionSetupData) (espMntDir string, unmount func() error, err error) {
+		mountVolsCalls++
+		return espDir, func() error { return nil }, nil
+	})
+	s.AddCleanup(restore)
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	// Mock saving of traits
+	saveStorageTraitsCalls := 0
+	restore = devicestate.MockInstallSaveStorageTraits(func(model gadget.Model, allLaidOutVols map[string]*gadget.LaidOutVolume, encryptSetupData *install.EncryptionSetupData) error {
+		saveStorageTraitsCalls++
+		return nil
+	})
+	s.AddCleanup(restore)
+
+	// Insert encryption data when enabled
+	if opts.encrypted {
+		restore = devicestate.MockEncryptionSetupDataInCache(s.state, label)
+		s.AddCleanup(restore)
+	}
+
+	// Create change
+	chg := s.state.NewChange("install-step-finish", "finish setup of run system")
+	finishTask := s.state.NewTask("install-finish", "install API finish step")
+	finishTask.Set("system-label", label)
+	finishTask.Set("on-volumes", ginfo.Volumes)
+
+	chg.AddTask(finishTask)
+
+	// now let the change run - some checks will happen in the mocked functions
+	s.state.Unlock()
+	defer s.state.Lock()
+
+	s.settle(c)
+
+	s.state.Lock()
+	defer s.state.Unlock()
+	c.Check(chg.Err(), IsNil)
+
+	// Checks now
+	kernelDir := filepath.Join(dirs.SnapRunDir, "snap-content/kernel")
+	c.Check(mountCmd.Calls(), DeepEquals, [][]string{
+		{"systemd-mount", gadgetSnapPath, gadgetDir},
+		{"systemd-mount", kernelSnapPath, kernelDir},
+		{"systemd-mount", "--umount", gadgetDir},
+		{"systemd-mount", "--umount", kernelDir},
+	})
+	c.Check(writeContentCalls, Equals, 1)
+	c.Check(mountVolsCalls, Equals, 1)
+	c.Check(saveStorageTraitsCalls, Equals, 1)
+
+	expectedFiles := []string{
+		filepath.Join(espDir, "EFI/ubuntu/grub.cfg"),
+		filepath.Join(espDir, "EFI/ubuntu/grubenv"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-boot/EFI/ubuntu/grub.cfg"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-boot/EFI/ubuntu/grubenv"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-boot/EFI/ubuntu/pc-kernel_1.snap/kernel.efi"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-boot/EFI/ubuntu/kernel.efi"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-boot/device/model"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-data/var/lib/snapd/modeenv"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-data/var/lib/snapd/snaps/core22_1.snap"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-data/var/lib/snapd/snaps/pc_1.snap"),
+		filepath.Join(dirs.RunDir, "mnt/ubuntu-data/var/lib/snapd/snaps/pc-kernel_1.snap"),
+	}
+	for _, f := range expectedFiles {
+		_, err := ioutil.ReadFile(f)
```

Comment:
(nitpick) this could use `c.Check(f, testutil.FilePresent)`

Reply:
Nice one, changed now